l3kernel/l3tl.dtx

```diff
@@ -3146,26 +3235,28 @@
 % \begin{macro}[EXP,pTF]{\tl_if_head_is_space:n}
 % \begin{macro}[EXP]{\@@_if_head_is_space:w}
 %   The auxiliary's argument is all that is before the first explicit
-%   space in |?#1?~|.  If that is a single~|?| the test yields
-%   \texttt{true}.  Otherwise, that is more than one token, and the test
-%   yields \texttt{false}.  The work is done within braces (with an
+%   space in |\prg_do_nothing:#1?~|.  If that is a single~|\prg_do_nothing:| the
+%   test yields \texttt{true}.  Otherwise, that is more than one token, and the
+%   test yields \texttt{false}.  The work is done within braces (with an
 %   |\if_false: { \fi: ... }| construction) both to hide potential
 %   alignment tab characters from \TeX{} in a table, and to allow for
-%   removing what remains of the token list after its first space.  The
-%   \cs{exp:w} and \cs{exp_end:} ensure that the result of a
-%   single step of expansion directly yields a balanced token list (no
-%   trailing closing brace).
+%   removing what remains of the token list after its first space.  The use of
+%   \cs{if:w} ensures that the result of a single step of expansion directly
+%   yields a balanced token list (no trailing closing brace).
 %    \begin{macrocode}
 \prg_new_conditional:Npnn \tl_if_head_is_space:n #1 { p , T , F , TF }
   {
-    \exp:w \if_false: { \fi:
-      \@@_if_head_is_space:w ? #1 ? ~ }
+    \if:w
+        \if_false: { \fi: \@@_if_head_is_space:w \prg_do_nothing: #1 ? ~ }
+        \scan_stop: \scan_stop:
+      \prg_return_true:
+    \else:
+      \prg_return_false:
+    \fi:
   }
 \cs_new:Npn \@@_if_head_is_space:w #1 ~
   {
-    \tl_if_empty:oTF { \use_none:n #1 }
-      { \exp_after:wN \exp_end: \exp_after:wN \prg_return_true: }
-      { \exp_after:wN \exp_end: \exp_after:wN \prg_return_false: }
+    \tl_if_empty:oF {#1} { f }
```

Comment:
Perhaps do `\@@_if_empty_if:o {#1} f \fi:`, and perhaps even the `\exp_args:Nno \use:n` construction you used a few times?  I wonder if we should simply define `\cs_new:Npo` and sprinkle that throughout most of `expl3`.

Reply:
It would have to be the reverse behaviour, so I would've to put `\exp_after:wN \reverse_if:N \@@_if_empty_if:o {#1} f \fi:` there, but that doesn't seem like a problem.

Yes the `\exp_args:Nno \use:n` construct feels strange when there could be a `\cs_new:Npo` variant.

Reply:
Indeed, I messed up the logic.  No need for `\reverse_if:N`, just use `\else:`, so do `\@@_if_empty_if:o {#1} \else: f \fi:`.

Reply:
`\reverse_if:N` is faster than `\else:`, hence I'd actually go down that route.

Reply:
Ok, there is some strangeness happening in `m3str002.lvt`. With `\else:` it works whereas with `\reverse_if:N` it fails, and I have no idea why.